pkg/controller/endpointslice/reconciler.go

```diff
@@ -251,6 +258,32 @@ func (r *reconciler) reconcileByAddressType(service *corev1.Service, pods []*cor
 	return r.finalize(service, slicesToCreate, slicesToUpdate, slicesToDelete, triggerTime)
 }
 
+// placeholderSliceEqual returns if two placeholder slices are equal.
+// It ignores many of the fields in ObjectMeta that are not expected to match.
+func placeholderSliceEqual(x, y *discovery.EndpointSlice) bool {
+	if x == nil || y == nil {
+		return x == y
+	}
+
+	// Override fields we don't care about in ObjectMeta so that we can use DeepEqual
+	xCopy := x.DeepCopy()
+	yCopy := y.DeepCopy()
+	xCopy.ObjectMeta = metav1.ObjectMeta{
+		Labels:          xCopy.ObjectMeta.Labels,
+		OwnerReferences: xCopy.OwnerReferences,
+		Namespace:       xCopy.Namespace,
+	}
+	yCopy.ObjectMeta = metav1.ObjectMeta{
+		Labels:          yCopy.ObjectMeta.Labels,
+		OwnerReferences: yCopy.OwnerReferences,
+		Namespace:       yCopy.Namespace,
+	}
+
+	result := apiequality.Semantic.DeepEqual(xCopy, yCopy)
+
+	return result
+}
+
```

Comment:
I will try to do a small benchmark to compare them, if nobody beats me

Reply:
Ran a quick bench between the two and the conversion func method is definitely more efficient.  Since we can make a bunch of assumptions about what the placeholder slice should have we can save some comparisons.  The downside of this is that if we ever update the desired placeholder slice logic we have to update this func as well.

https://gist.github.com/spencerhance/a607871b5490a5ff6b2bfe52571eb496

Reply:
Why does it need to be deep-copy?  In this case a shallow copy seems perfect, unless I am missing something?

```
	// Make a shallow copy (cheap) and wipe all metadata fields except the
	// ones we actually care about, so that we can use DeepEqual.
	xCopy := x // shallow copy
	yCopy := y // shallow copy
	xCopy.ObjectMeta = metav1.ObjectMeta{
		Labels:          xCopy.ObjectMeta.Labels,
		OwnerReferences: xCopy.OwnerReferences,
		Namespace:       xCopy.Namespace,
	}
	yCopy.ObjectMeta = metav1.ObjectMeta{
		Labels:          yCopy.ObjectMeta.Labels,
		OwnerReferences: yCopy.OwnerReferences,
		Namespace:       yCopy.Namespace,
	}
```

Reply:
+1 for shallow copy

That said I think we can do much better than that by defining our customer DeepEqual function.
Given we create placeholder slice, we can probably just focus on checking only the couple fields of ObjectMeta.
And we can define our file-level Equality that will just overwrite comparing of ObjectMeta to limit to those fields:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/api/equality/semantic.go#L29

And with that we won't even need shallow copies.

Reply:
Good call, the shallow copy also works here.  I pushed an update with the shallow copy but I can also use the conversion func method.  How does this look?

```
	compare := conversion.EqualitiesOrDie(
		func(a, b metav1.OwnerReference) bool {
			return a.String() == b.String()
		},
		func(a, b metav1.ObjectMeta) bool {
			if a.Namespace != b.Namespace {
				return false
			}
			for k, v := range a.Labels {
				if b.Labels[k] != v {
					return false
				}
			}
			for k, v := range b.Labels {
				if a.Labels[k] != v {
					return false
				}
			}
			return true
		},
```